src/ts/AdExperienceTrackingPlugin.ts

```diff
@@ -0,0 +1,132 @@
+    const ad = event.ad as LinearAd;
+
+    if (!adData) {
+      this.logger.consoleLog(
+        'For proper adExperience tracking please use player version >= \'8.3.0\'.',
+        Conviva.SystemSettings.LogLevel.WARNING,
+      );
+    }
+
+    // Create a new ContentMetadata object for ad.
+    let adMetadata = new Conviva.ContentMetadata();
+    adMetadata.assetName = adData && adData.adTitle || AdExperienceTrackingPlugin.UNKNOWN_VALUE;
+    adMetadata.streamUrl = ad.mediaFileUrl;
+    adMetadata.duration = ad.duration; // Ad Duration In seconds
+
+    const isYospaceUsed = this.player.ads.getModuleInfo().name.includes('yospace');
+    const adTechnology = isYospaceUsed ?
+      Conviva.Client.AdTechnology.SERVER_SIDE :
+      Conviva.Client.AdTechnology.CLIENT_SIDE;
+
+    // custom takes a javascript object as an argument
+    adMetadata.custom = {
+      // Required
+      'c3.ad.technology': String(adTechnology),
+      'c3.ad.id': ad.id || AdExperienceTrackingPlugin.UNKNOWN_VALUE,
+      'c3.ad.system': adData && adData.adSystem && adData.adSystem.name || AdExperienceTrackingPlugin.UNKNOWN_VALUE,
+      'c3.ad.position': String(this.currentAdBreakPosition),
+      'c3.ad.type': AdExperienceTrackingPlugin.UNKNOWN_VALUE,
+      'c3.ad.mediaFileApiFramework': adData && adData.apiFramework || AdExperienceTrackingPlugin.UNKNOWN_VALUE,
+      'c3.ad.adStitcher': isYospaceUsed ? 'Yospace' : AdExperienceTrackingPlugin.UNKNOWN_VALUE,
+    };
+
+    const addValueIfPresent = (key: string, value: string) => {
+      if (value) {
+        adMetadata.custom[key] = value;
+      }
+    };
+
+    addValueIfPresent('c3.ad.creativeId', adData && adData.creative && adData.creative.adId);
+    addValueIfPresent(
+      'c3.ad.creativeName',
+      adData && adData.creative && adData.creative.universalAdId && adData.creative.universalAdId.value,
+    );
+    addValueIfPresent('c3.ad.breakId', this.currentAdBreak.id);
+    addValueIfPresent('c3.ad.advertiser', adData && adData.advertiser && adData.advertiser.name);
+    addValueIfPresent('c3.ad.advertiserId', adData && adData.advertiser && adData.advertiser.id);
+
+    this.adSessionKey = this.client.createAdSession(this.contentSessionKey, adMetadata);
+    this.adPlayerStateManager = this.client.getPlayerStateManager();
+
+    // Client object is obtained from the main video content
+    this.client.attachPlayer(this.adSessionKey, this.adPlayerStateManager);
+
+    this.adPlayerStateManager.setPlayerState(Conviva.PlayerStateManager.PlayerState.PLAYING);
+
+    const adBitrate = adData && adData.bitrate;
+    this.reportBitrate(adBitrate);
+  }
+
+  private reportBitrate(bitrate: number) {
+    if (!bitrate) {
+      return;
+    }
+
+    let kbpsBitrate = bitrate;
+    // Since the unit isn't defined for adData.bitrate (neither in the Vast Manifest) we assume that values greater
+    // than 10_000 are in bps and convert them to kbps
+    if (bitrate >= 10_000) {
```

Comment:
Guess there is no way to make this future proof? Hope nobody will get the idea to serve 10mbit ads 🙂